sdk/eventhub/Azure.Messaging.EventHubs/src/Core/TransportProducerPool.cs

```diff
@@ -0,0 +1,263 @@
+        ///   are in use and those that can be closed.
+        /// </summary>
+        ///
+        private Timer ExpirationTimer { get; }
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="TransportProducerPool"/> class.
+        /// </summary>
+        ///
+        public TransportProducerPool()
+        {
+            ExpirationTimer = new Timer(PerformExpiration(),
+                                        null,
+                                        TimeSpan.FromMinutes(PerformExpirationPeriodAsMinutes),
+                                        TimeSpan.FromMinutes(PerformExpirationPeriodAsMinutes));
+        }
+
+        /// <summary>
+        ///   It returns a <see cref="PooledProducer"/> corresponding to a given partition
+        ///   and it extends its sliding expiration.
+        /// </summary>
+        ///
+        /// <param name="partitionId">The zero-based unique identifier of a partition associated with the Event Hub.</param>
+        /// <param name="connection">The <see cref="EventHubConnection"/> to use as the basis for delegation of client-type operations.</param>
+        /// <param name="retryPolicy">The <see cref="EventHubsRetryPolicy"/> to use as the basis for retrieving the information.</param>
+        ///
+        /// <returns>A <see cref="PooledProducer"/> corresponding to the partitionId passed in as parameter.</returns>
+        ///
+        /// <remarks>
+        ///   There is a slight probability that the returned producer may be closed at any time
+        ///   after it is returned and the caller may want to handle that scenario.
+        /// </remarks>
+        ///
+        public PooledProducer GetPartitionProducer(string partitionId,
+                                                   EventHubConnection connection,
+                                                   EventHubsRetryPolicy retryPolicy)
+        {
+            Argument.AssertNotNull(partitionId, nameof(partitionId));
+            Argument.AssertNotNull(connection, nameof(connection));
+            Argument.AssertNotNull(retryPolicy, nameof(retryPolicy));
+
+            var identifier = Guid.NewGuid().ToString();
+
+            var item = Pool.GetOrAdd(partitionId, id => new PoolItem(connection.CreateTransportProducer(id, retryPolicy)));
+
+            // A race condition at this point may end with an CloseAsync called on
+            // the returned PoolItem if it had expired. The probability is very low and
+            // possible exceptions should be handled by the invoking methods.
+
+            if (!item.ActiveInstances.TryAdd(identifier, 0) || item.Producer.IsClosed)
+            {
+                identifier = Guid.NewGuid().ToString();
+                item = Pool.GetOrAdd(partitionId, id => new PoolItem(connection.CreateTransportProducer(id, retryPolicy)));
+                item.ActiveInstances.TryAdd(identifier, 0);
+            }
```

Comment:
I will admit I am not completely familiar with the topic and the numbers involved but I would be curious to discuss it better as we left it behind.

I am aware the key is hashed and the hash is used to compute the bucket number and the lock number in ```ConcurrentDictionary<TKey, TValue>```. Still ```TryAdd``` may return ```false``` only when the keys are the same.

Reply:
There's a very, very, very slim chance that it could collide.  Not something that we should ever see in normal use, but as an added guard for a clause that we already need, it's the right thing to do, IMHO.